docs/source/core/data-structures/zend_constant.rst

```diff
@@ -0,0 +1,64 @@
+###############
+ zend_constant
+###############
+
+PHP constants (referring to non-class constants) are stored in a dedicated structure
+``zend_constant``, which holds both the value of the constant and details for using it.
+
+************
+ definition
+************
+
+.. code:: c
+
+   typedef struct _zend_constant {
+       zval value;
+       zend_string *name;
+       zend_string *filename;
+   } zend_constant;
+
+The ``value`` field stores both the value itself and some metadata. The ``name`` and ``filename``
+store the name of the constant and the name of the file in which it was defined.
+
+*******
+ value
+*******
+
+The value of the constant is stored in the :doc:`./zval` ``value``. However, since the ``zval``
+structure has extra space, for constants this is used to store both the number of the module that
+the constant was defined in, and a combination of the flags that affect the usage of the constant.
+
+This extra information is placed in the ``uint32_t`` field ``value.u2.constant_flags``.
+
+The bottom 16 bits are used to hold flags about the constant
+
+.. code:: c
+
+   #define CONST_PERSISTENT     (1<<0) /* Persistent */
+   #define CONST_NO_FILE_CACHE  (1<<1) /* Can't be saved in file cache */
```

Comment:
This apparently only applies to PHP_SAPI and PHP_BINARY. Maybe we can expand on why that is, i.e. to prevent incorrect values (values that can change between environments while allowing to reuse cache files) from being inlined into the file cache.